File: Trees/C_Cpp/lowest_common_ancestor.c
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

struct TreeNode* lowestCommonAncestor(struct TreeNode* root, struct TreeNode* p, struct TreeNode* q) {

    if (!root || root == p || root == q) {
        return root;
    }
    struct TreeNode* left = lowestCommonAncestor(root->left, p, q);
    struct TreeNode* right = lowestCommonAncestor(root->right, p, q);

    if (left && right) {
        return root;
    }
    return left ? left : right;
}
/*
* alternatively
 struct TreeNode* cur = root;
 while (cur) {
   if ((p->val < cur->val) && (q->val < cur->val)) {
       cur = cur->left;
   } else if ((p->val > cur->val) && (q->val > cur->val)) {
       cur = cur->right;
   } else {
       return cur;
   }

 }
*/

File: Trees/Python/lowest_common_ancestor.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        if not root or root == p or root == q:
            return root
        left = self.lowestCommonAncestor(root.left, p ,q)
        right = self.lowestCommonAncestor(root.right, p, q)

        if left and right:
            return root
        return left if left else right


"""
alternatively
class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        cur = root
        while cur:
            if (p.val > cur.val and q.val > cur.val):
                cur = cur.right
            elif (p.val < cur.val and q.val < cur.val):
                cur = cur.left
            else:
                return cur

"""

